src/main/java/org/dswarm/xsd2jsonschema/JsonSchemaParser.java

```diff
@@ -54,155 +57,234 @@
 
 public class JsonSchemaParser {
 
-	private final XSOMParser parser = new XSOMParser();
+	private static final String HASH        = "#";
+	private static final String HTTP_PREFIX = "http://";
+	private static final String SLASH       = "/";
+	private static final String AT          = "@";
+	private static final String WILDCARD    = "wildcard";
+	private static final String NULL        = "null";
+
+	//private static final XSImplementation impl         = (XSImplementation) new XSImplementationImpl();
+
+	private static XSImplementation impl;
+
+	static {
+
+		System.setProperty(DOMImplementationRegistry.PROPERTY, "com.sun.org.apache.xerces.internal.dom.DOMXSImplementationSourceImpl");
+
+		try {
+
+			final DOMImplementationRegistry registry = DOMImplementationRegistry.newInstance();
+
+			impl = (XSImplementation) registry.getDOMImplementation("XS-Loader");
+		} catch (ClassNotFoundException | InstantiationException | IllegalAccessException e) {
+
+			e.printStackTrace();
+		}
+	}
+
+	private static final XSLoader schemaLoader = impl.createXSLoader(null);
+
+	private XSModel model = null;
 
 	private List<JSElement> iterateParticle(final XSParticle particle) {
 
 		final XSTerm term = particle.getTerm();
 
-		if (term.isModelGroup()) {
+		if (term instanceof XSModelGroup) {
 
-			final XSModelGroup modelGroup = term.asModelGroup();
+			final XSModelGroup modelGroup = (XSModelGroup) term;
 			return iterateModelGroup(modelGroup);
-		} else if (term.isModelGroupDecl()) {
+		}
+
+		// TODO: dunno how to replace this? -> we can cast it to XSModelGroup but maybe it should ba a XSModelGroupDefinition and has something to do with substition groups
+		/*else if (term.isModelGroupDecl()) {
 
 			final XSModelGroupDecl xsModelGroupDecl = term.asModelGroupDecl();
 			return iterateModelGroup(xsModelGroupDecl.getModelGroup());
 		}
+		*/
 
 		final ArrayList<JSElement> jsElements = new ArrayList<>(1);
-		jsElements.add(iterateSingleParticle(particle));
+		final Optional<JSElement> optionalJSElement = iterateSingleParticle(particle);
+
+		if (optionalJSElement.isPresent()) {
+
+			jsElements.add(optionalJSElement.get());
+		}
 
 		return jsElements;
 	}
 
-	private JSElement iterateSingleParticle(final XSParticle particle) {
+	private Optional<JSElement> iterateSingleParticle(final XSParticle particle) {
 
 		final XSTerm term = particle.getTerm();
-		if (term.isElementDecl()) {
+		if (term instanceof XSElementDeclaration) {
+
+			final XSElementDeclaration xsElementDecl = (XSElementDeclaration) term;
 
-			final XSElementDecl xsElementDecl = term.asElementDecl();
+			final Optional<JSElement> optionalElement = iterateElement(xsElementDecl);
 
-			final JSElement element = iterateElement(xsElementDecl);
+			if (!optionalElement.isPresent()) {
 
-			return particle.isRepeated() ? new JSArray(element) : element;
-		} else if (term.isModelGroupDecl()) {
+				return Optional.empty();
```

Comment:
Since we're now in java8 land, the isPresent check can be replaced by this line

```java
return optionalElement.map(elem -> isRepeated(particle) ? new JSArray(elem) : elem);
```